lib/kafka_ex/server_kayrock.ex

```diff
@@ -0,0 +1,946 @@
+    end
+
+    brokers =
+      state.bootstrap_uris
+      |> Enum.with_index()
+      |> Enum.into(%{}, fn {{host, port}, ix} ->
+        {ix + 1,
+         %Broker{
+           host: host,
+           port: port,
+           socket:
+             NetworkClient.create_socket(
+               host,
+               port,
+               state.ssl_options,
+               state.use_ssl
+             )
+         }}
+      end)
+
+    check_brokers_sockets!(brokers)
+
+    state = %{state | cluster_metadata: %ClusterMetadata{brokers: brokers}}
+
+    {ok_or_err, api_versions, state} = get_api_versions(state)
+
+    if ok_or_err == :error do
+      sleep_for_reconnect()
+      raise "Brokers sockets are closed"
+    end
+
+    :no_error = Kayrock.ErrorCode.code_to_atom(api_versions.error_code)
+
+    state = State.ingest_api_versions(state, api_versions)
+
+    state =
+      try do
+        update_metadata(state)
+      rescue
+        e ->
+          sleep_for_reconnect()
+          Kernel.reraise(e, System.stacktrace())
+      end
+
+    {:ok, _} =
+      :timer.send_interval(state.metadata_update_interval, :update_metadata)
+
+    {:ok, state}
+  end
+
+  @impl true
+  def handle_call(:cluster_metadata, _from, state) do
+    {:reply, {:ok, state.cluster_metadata}, state}
+  end
+
+  def handle_call(:correlation_id, _from, state) do
+    {:reply, {:ok, state.correlation_id}, state}
+  end
+
+  def handle_call(:update_metadata, _from, state) do
+    updated_state = update_metadata(state)
+    {:reply, {:ok, updated_state.cluster_metadata}, updated_state}
+  end
+
+  def handle_call(
+        {:set_consumer_group_for_auto_commit, consumer_group},
+        _from,
+        state
+      ) do
+    if KafkaEx.valid_consumer_group?(consumer_group) do
+      {:reply, :ok, %{state | consumer_group_for_auto_commit: consumer_group}}
+    else
+      {:reply, {:error, :invalid_consumer_group}, state}
+    end
+  end
+
+  def handle_call({:topic_metadata, topics, allow_topic_creation}, _from, state) do
+    allow_auto_topic_creation = state.allow_auto_topic_creation
+
+    updated_state =
+      update_metadata(
+        %{state | allow_auto_topic_creation: allow_topic_creation},
+        topics
+      )
+
+    topic_metadata = State.topics_metadata(updated_state, topics)
+
+    {:reply, {:ok, topic_metadata},
+     %{updated_state | allow_auto_topic_creation: allow_auto_topic_creation}}
+  end
+
+  def handle_call({:kayrock_request, request, node_selector}, _from, state) do
+    {response, updated_state} =
+      kayrock_network_request(request, node_selector, state)
+
+    {:reply, response, updated_state}
+  end
+
+  def handle_call({:metadata, topic}, _from, state) do
+    updated_state = update_metadata(state, [topic])
+
+    {:reply, Adapter.metadata_response(updated_state.cluster_metadata),
+     updated_state}
+  end
+
+  ######################################################################
+  # compatibility handle_call messages
```

Comment:
I think it would be interesting to have a seperate GenServer for the compatbilitiy with the existing API, and have that server call a separate GenServer implementing the new API (by sending `:kayrock_request` calls). It would help make sure that there's a clear seperation between code that's there to maintain the old API and the code implementing the new API.

Reply:
Hm, I think I follow.  You would have a GenServer implementation that's compatible with the old API that does all of the adapting work but then just relays to the new client for the work?  That would definitely keep things separated but I'm worried that adding that extra server layer will complicate things.  With the implementation here I was going for a drop-in replacement, especially with the ConsumerGroup code which is much newer and cleaner than the old KafkaEx API and I hope to keep most of that untouched.

Reply:
I'm not sure I understand everything in your answer :/. 

>  especially with the ConsumerGroup code which is much newer and cleaner than the old KafkaEx API and I hope to keep most of that untouched.

You mean you want to be able to keep the new code you wrote for when you'll write the new API ?

Anyway, here's what I'm worried about, It's probably better to start with that: I'm worried that `ServerKayrock` will contain both:
1. code that is needed in the new, future `KafkaEx.Server`
2. code that is there to implement the legacy `KafkaEx.Server` API.

That worries me because:
- it could taint the logic of the code that needs to be kept, because someone will tweak it to "make it work the the legacy API"
- it could make it hard to discern what must be removed once the legacy API is removed.

My suggestion comes from that worry. To prevent leakage from the legacy API to the new API, I figured having a proxy process would help -- all the "good new code" we want to keep would be kept separately from the code that's only needed to maintain compatibility.

Alternative suggestion: 
1. extract all the logic out of the GenServer. 
2. make two GenServers, one for compatibility and one for the new API
3. use the common, extracted logic in both

Maybe my worries are not justified -- maybe I just don't understand well enough how it works ?

Reply:
Ah I see.  Most of the logic in `ServerKayrock` is intended to be the new logic that will stick around.  It's based heavily on the existing `KafkaEx.Server` logic because that's sort of battle-tested at this point, but with some refactoring to make it more maintainable and also trying to keep as much of the request processing logic outside of the server (this tends to be what makes keeping up with the Kafka API difficult).

I'll try to think of some way to make the compatibility code more self-contained so that it's easier to extract later.

Reply:
I moved all of the compatibility `handle_call` clauses into a mixin module - does that work?  I don't want to expend _too_ much energy or make the code a lot more complex to avoid something that we should catch in code review during the transition period.  I'm open to discussion on that, though.